File: lcof-python/剑指offer/1_ListNode/3_链表内指定区间反转.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
#
# 代码中的类名、方法名、参数名已经指定，请勿修改，直接返回方法规定的值即可
#
# 
# @param head ListNode类 
# @param m int整型 
# @param n int整型 
# @return ListNode类
#
# 非剑指offer
class Solution:
    def reverseBetween(self , head: ListNode, m: int, n: int) -> ListNode:
        # write code here
        if not head:
            return None
        if m == n:
            return head
        count = 1
        temp1 = head
        pre = None
        after = head
        while head:
            # 保存断掉的节点
            # 1 连接 4； 4 -》 head_connect.next = first
            # 2 连接 5； 5 -》 end.next = tail_connect
            #
            if count == m:
                head_connect = pre
                end = after
            if count == n:
                tail_connect = after.next
                first = after
            # 可能会导致断链 after == head
            # 3 个元素 调整2次链接
            if count >= m and count <=n:
                temp = after.next
                after.next = pre
                pre = after
                after = temp
                head = after
            else:
                pre = after
                after = after.next
                head = head.next
            count += 1
        # m == 1 1 链接 5 
        # n == 5 2 链接 Null
        if m == 1:
            end.next = tail_connect
            temp1 = first
        # elif (count - 1) == n:
        #     head_connect.next = first
        #     end.next = None
        # elif m == 1 and (count - 1) == n:
        #     pass
        else:
            end.next = tail_connect
            head_connect.next = first
        return temp1

File: lcof-cpp/offer/1_ListNode/3_链表内指定区间反转.cpp
/**
 * struct ListNode {
 *	int val;
 *	struct ListNode *next;
 *	ListNode(int x) : val(x), next(nullptr) {}
 * };
 */
#include <memory>
class Solution
{
public:
    /**
     * 代码中的类名、方法名、参数名已经指定，请勿修改，直接返回方法规定的值即可
     *
     *
     * @param head ListNode类
     * @param m int整型
     * @param n int整型
     * @return ListNode类
     */
    ListNode *reverseBetween(ListNode *head, int m, int n)
    {
        if (!head)
            return nullptr;
        auto dummyHead = new ListNode(-1);
        dummyHead->next = head;
        auto pre = dummyHead; // 记录要更改区间的前一个节点
        for (int i = 0; i < m - 1; ++i)
            pre = pre->next;
        // 类似插入排序 将第一个一次一次 反转到最后， 但是要把最后的向前反转 所以每一次反转都要把cur_next节点向最开始的位置反转
        auto cur = pre->next; // 要反转到最后的节点
        ListNode *cur_next = nullptr;
        for (int i = 0; i < n - m; ++i)
        {
            // 更新cur_next结点
            cur_next = cur->next;
            cur->next = cur_next->next; // 将cur->next连接指向下一个节点
            cur_next->next = pre->next; // 反转到第一个结点 类似LRUcache的moveToFront
            pre->next = cur_next;
        }
        return dummyHead->next;
    }
};